src/julia_internal.h

```diff
@@ -590,7 +590,108 @@ jl_tupletype_t *arg_type_tuple(jl_value_t *arg1, jl_value_t **args, size_t nargs
 
 int jl_has_meta(jl_array_t *body, jl_sym_t *sym);
 
-// backtraces
+//--------------------------------------------------
+// Backtraces
+
+// Backtrace buffers:
+//
+// A backtrace buffer conceptually contains a stack of instruction pointers
+// ordered from the inner-most frame to the outermost. We store them in a
+// special raw format for two reasons:
+//
+//   * Efficiency: Every `throw()` must populate the trace so it must be as
+//     efficient as possible.
+//   * Signal safety: For signal-based exceptions such as StackOverflowError
+//     the trace buffer needs to be filled from a signal handler where most
+//     operations are not allowed (including malloc) so we choose a flat
+//     preallocated buffer.
+//
+// The raw buffer layout contains "frame entries" composed of one or several
+// `jl_bt_element_t` values. From the point of view of the GC, an entry is
+// either:
+//
+// 1. A single instruction pointer to native code, not GC-managed.
+// 2. An "extended entry": a mixture of raw data and pointers to julia objects
+//    which must be treated as GC roots.
+//
+// An extended entry `e` is made up of several jl_bt_element_t values:
+//
+//   e[0]  JL_BT_NON_PTR_ENTRY  - Special marker to distinguish extended entries
+//   e[1]  tags                 - A bit packed uintptr_t containing a tag and
+//                                the number of GC- managed and non-managed values
+//   e[2+j]                     - GC managed data
+//   e[2+ngc+i]                 - Non-GC-managed data
+//
+// The format of `tags` is, from LSB to MSB:
+//   0:2     ngc     Number of GC-managed pointers for this frame entry
+//   3:5     nptr    Number of non-GC-managed buffer elements
+//   6:9     tag     Entry type
+//   10:...  header  Entry-specific header data
+typedef struct _jl_bt_element_t {
+    union {
+        uintptr_t   uintptr; // Metadata or native instruction ptr
+        jl_value_t* jlvalue; // Pointer to GC-managed value
+    };
+} jl_bt_element_t;
+
+#define JL_BT_NON_PTR_ENTRY (((uintptr_t)0)-1)
+// Maximum size for an extended backtrace entry (likely significantly larger
+// than the actual size of 3-4 for an interpreter frame)
+#define JL_BT_MAX_ENTRY_SIZE 16
+
+STATIC_INLINE int jl_bt_is_native(jl_bt_element_t *bt_entry) JL_NOTSAFEPOINT
+{
+    return bt_entry[0].uintptr != JL_BT_NON_PTR_ENTRY;
+}
+
+// Extended backtrace entry header packing; the bit packing is done manually
+// for precise layout control for interop with julia side.
+STATIC_INLINE uintptr_t jl_bt_pack_extended_entry(int ngc, int nptr,
```

Comment:
This name implies it creates an entire `jl_bt_element_t`, not just the `tags`. Maybe `jl_bt_extended_entry_descriptor`?